File: Coursera/DataStructures/week4/substring_equality.cpp
#include <iostream>
#include <vector>
#include <random>
using namespace std;

class Solver {
	string s;
    vector<unsigned long long> h1;
    vector<unsigned long long> h2;
    vector<unsigned long long> powsM1;
    vector<unsigned long long> powsM2;
     unsigned long long m1 = 1000000007;
     unsigned long long m2 = 1000000009;
     unsigned long long x = 4;
public:
	Solver(string s) : s(s) {
		// initialization, precalculation

        this->powsM1 = vector<unsigned long long>(s.size() + 1);
        this->powsM2 = vector<unsigned long long>(s.size() + 1);
        this->h1 = vector< unsigned long long>(s.size() + 1);
        this->h2 = vector< unsigned long long>(s.size() + 1);
        this->powsM1[0] = 1;
        this->powsM2[0] = 1;
        this->h1[0] = 0;
        this->h2[0] = 0;

        for (int i = 1; i <= s.size(); i++){
            this->powsM1[i] = (this->powsM1[i - 1] * this->x) % this->m1;
            this->powsM2[i] = (this->powsM2[i - 1] * this->x) % this->m2;
            this->h1[i] = (this->h1[i - 1] + (s[i - 1] - 'a' + 1) * this->powsM1[i - 1]) % this->m1;
            this->h2[i] = (this->h2[i - 1] + (s[i - 1] - 'a' + 1) * this->powsM2[i - 1]) % this->m2;
        }
	}

	bool ask(int a, int b, int l) {
         long long lH1 = ((this->h1[a + l] - this->h1[a]) * this->powsM1[b - a]) % this->m1;
         long long rH1 = ((this->h1[b + l] - this->h1[b])) % this->m1;
         long long lH2 = ((this->h2[a + l] - this->h2[a]) * this->powsM2[b - a]) % this->m2;
         long long rH2 = ((this->h2[b + l] - this->h2[b])) % this->m2;

         //cout << lH1 << " " << rH1 << " " << lH2 << " " << rH2 << " ";

		return lH1 == rH1 &&
		    lH2 == rH2;
	}
};

int main() {
	ios_base::sync_with_stdio(0), cin.tie(0);

	string s;
	int q;
	cin >> s >> q;
	Solver solver(s);
	for (int i = 0; i < q; i++) {
		int a, b, l;
		cin >> a >> b >> l;
		cout << (solver.ask(a, b, l) ? "Yes\n" : "No\n");
	}
}

File: Coursera/DataStructures/week4/substring_equality.py
# python3
from random import randint
import sys
import numpy as np

class Solver:
	def __init__(self, s):
		self.s = s
		self.m1 = 10 ** 9 + 7
		self.m2 = 10 ** 9 + 9
		self.x = randint(1, 10 ** 9)
		self.h1 = np.zeros(len(self.s))
		self.h2 = np.zeros(len(self.s))
		self.powsM1 = np.ones(len(self.s))
		self.powsM2 = np.ones(len(self.s))

		for i in range(1, len(self.s)):
			self.powsM1[i] = self.powsM1[i - 1] * self.x % self.m1
			self.powsM2[i] = self.powsM2[i - 1] * self.x % self.m2
			self.h1[i] = (self.h1[i - 1] + ord(self.s[i - 1]) * self.powsM1[i - 1]) % self.m1
			self.h2[i] = (self.h2[i - 1] + ord(self.s[i - 1]) * self.powsM2[i - 1]) % self.m2

	def ask(self, a, b, l):
		firstHashH1 = (self.h1[a + l] - self.h1[a]) / self.powsM1[a] % self.m1
		secondHashH1 = (self.h1[b + l] - self.h1[b]) / self.powsM1[b] % self.m1
		firstHashH2 = (self.h2[a + l] - self.h2[a]) / self.powsM1[a] % self.m2
		secondHashH2 = (self.h2[b + l] - self.h2[b]) / self.powsM1[b] % self.m2

		print( firstHashH1, " ", secondHashH1, " ",firstHashH2, " ", secondHashH2)

		return firstHashH1 == secondHashH1 and firstHashH2 == secondHashH2

s = sys.stdin.readline()
q = int(sys.stdin.readline())
solver = Solver(s)
for i in range(q):
	a, b, l = map(int, sys.stdin.readline().split())
	print("Yes" if solver.ask(a, b, l) else "No")
